share/completions/sudo.fish

```diff
@@ -2,54 +2,62 @@
 # Completion for sudo
 #
 
-# All these options should be valid for GNU and OSX sudo
-complete -c sudo -s A -d "Ask for password via the askpass or \$SSH_ASKPASS program"
-complete -c sudo -s C -d "Close all file descriptors greater or equal to the given number" -a "(seq 0 255)"
-complete -c sudo -s E -d "Preserve environment"
-complete -c sudo -s H -d "Set home"
-complete -c sudo -s K -d "Remove the credential timestamp entirely"
-complete -c sudo -s P -d "Preserve group vector"
-complete -c sudo -s S -d "Read password from stdin"
-complete -c sudo -s b -d "Run command in the background"
-complete -c sudo -s e -r -d "Edit"
-complete -c sudo -s g -a "(__fish_complete_groups)" -x -d "Run command as group"
-complete -c sudo -s h -n "__fish_no_arguments" -d "Display help and exit"
-complete -c sudo -s i -d "Run a login shell"
-complete -c sudo -s k -d "Reset or ignore the credential timestamp"
-complete -c sudo -s l -d "List the allowed and forbidden commands for the given user, or the full path to the given command if it is allowed"
-complete -c sudo -s n -d "Do not prompt for a password - if one is needed, fail"
-complete -c sudo -s p -d "Specify a custom password prompt"
-complete -c sudo -s s -d "Run the given command in a shell"
-complete -c sudo -s u -a "(__fish_complete_users)" -x -d "Run command as user"
-complete -c sudo -s v -n "__fish_no_arguments" -d "Validate the credentials, extending timeout"
+function __fish_sudo_print_remaining_args
+    set -l tokens (commandline -opc) (commandline -ct)
+    set -e tokens[1]
+    # These are all the options mentioned in the man page for Todd Miller's "sudo.ws" sudo (in that order).
+    # If any other implementation has different options, this should be harmless, since they shouldn't be used anyway.
+    set -l opts A/askpass b/background C/close-from= E/preserve-env=?
+    # Note that "-h" is both "--host" (which takes an option) and "--help" (which doesn't).
+    # But `-h` as `--help` only counts when it's the only argument (`sudo -h`),
+    # so any argument completion after that should take it as "--host".
+    set -a opts e/edit g/group= H/set-home h/host= '1-help'
+    set -a opts i/login K/remove-timestamp k/reset-timestamp l/list n/non-interactive
+    set -a opts P/preserve-groups p/prompt= S/stdin s/shell U/other-user=
+    set -a opts u/user= T/command-timeout= V/version v/validate
+    argparse -s $opts -- $tokens 2>/dev/null
+    # The remaining argv is the subcommand with all its options, which is what
+    # we want.
+    if test -n "$argv"
+        and not string match -qr '^-' $argv[1]
+        echo $argv
+        return 0
+    else
+        # Without this echo, complete won't get the -C option
+        echo
```

Comment:
It seems that an argument like `--option=(func_without_output)` would be just ignored. 
Any better ideas to deal with this?

Reply:
Ah, okay.

So, the issue this is almost certainly hitting is the cartesian product - see http://fishshell.com/docs/current/index.html#cartesian-product.

Let's say this function printed two lines - "a" and "b". `complete -C(__fish_sudo_print_remaining_args)` would then expand to `complete -Ca -Cb`, as if they were given as `complete -C{a,b}`.

If it printed nothing, it would expand to nothing. Printing an empty line instead would expand like `complete -C""`.

What you actually want here is to join all lines with a space, so you want

```fish
function __fish_complete_sudo_subcommand
    set -l args (__fish_sudo_print_remaining_args)
    complete -C"$args"
end
```